src/helper_gitlab_modify.sh

```diff
@@ -727,64 +746,63 @@ push_changes_to_gitlab() {
-        
+
         # If the GitLab repository exists
         if [ "$(gitlab_repo_exists_locally "$gitlab_repo_name")" == "FOUND" ]; then
-          
+
           # If the GitLab branch exists
-          
+
           local found_branch_name
 		  found_branch_name=$(get_current_gitlab_branch $gitlab_repo_name $gitlab_branch_name "GitLab")
           if [ "$found_branch_name" == "$gitlab_branch_name" ]; then
-          
+
             # If there exist differences in the files or folders in the branch (excluding the .git directory)
-            
+
             # Then copy the files and folders from the GitHub branch into the GitLab branch (excluding the .git directory)
             # That also deletes the files that exist in the GitLab branch that do not exist in the GitHub branch (excluding the .git directory)
             copy_github_files_and_folders_to_gitlab "$MIRROR_LOCATION/GitHub/$github_repo_name" "$MIRROR_LOCATION/GitLab/$github_repo_name"
-            
+
             # Then verify the checksum of the files and folders in the branches are identical (excluding the .git directory)
             local comparison_result
 			comparison_result="$(two_folders_are_identical_excluding_subdir $MIRROR_LOCATION/GitHub/$github_repo_name $MIRROR_LOCATION/GitLab/$github_repo_name .git)"
-            
+
             # Verify the files were correctly copied from GitHub branch to GitLab branch.
             if [ "$comparison_result" == "IDENTICAL" ]; then
               #echo "IDENTICAL"
-              
+
               # Get the path before executing the command (to verify it is restored correctly after).
               local pwd_before
 			  pwd_before="$PWD"
-              
+
               # TODO: Verify the changes were committed to GitLab correctly. (There are no remaining files to be added)
               #git status
               # TODO: Verify the changes were committed to GitLab correctly. (There commit message equals the sha)
               #git log
-              
+
               # Commit the changes to GitLab.
               cd "$MIRROR_LOCATION/GitLab/$github_repo_name" && git push --set-upstream origin "$gitlab_branch_name"
-              # cd "$MIRROR_LOCATION/GitLab/$github_repo_name" && git push
-			  # --set-upstream origin main
               cd ../../../..
-              
-              # Get the path after executing the command (to verify it is 
+
+              # Get the path after executing the command (to verify it is
 			  # restored correctly after).
               local pwd_after
 			  pwd_after="$PWD"
-              
+
               # Verify the current path is the same as it was when this function started.
               path_before_equals_path_after_command "$pwd_before" "$pwd_after"
-              
+
             else
-              echo "ERROR, the content in the GitHub branch is not exactly 
+              echo <<END "ERROR, the content in the GitHub branch is not exactly
```

Comment:
Nice work! Thanks for showing me how that's done!

Reply:
How can I verify if it works accordingly?

Reply:
There are (at least 2 ways): 
0. I/someone ensure(s) this repository works and make sure the CI automatically runs on your pull request. 
1. You run your own CI on this repo.
2. You run the tests locally (That should be(come) standard practice).
3. You write your own test for this function and run it locally.

I think currently option 2 or 3 is easiest. The instructions to do so are in the Readme. They are:  clone the repo,  `cd into the repo` and then you can install and run your own GitLab CI with:
```
chmod +x *.sh
./install_gitlab.sh -s -r
```
I did not generalise it properly yet, so I'd love to hear which errors you get so that I can work towards making it work on other operating systems (E.g. Windows) and other devices as well!

And then run the tests with:
```
rm -r test/libs/*
chmod +x *.sh
./install-bats-libs.sh
./test.sh
```
That runs all the tests. 


However, the easiest way is to clone it and just run the one test that you want to run, so instead of installing the GitLab server, you could just run:
```
rm -r test/libs/*
chmod +x *.sh
./install-bats-libs.sh
./test/libs/bats/bin/bats test/test_helper_gitlab_modify.bats
```
However, this particular test, requires the GitLab server to be installed and running. To only run the function you want to test e.g.  you could skip all other tests with:

```
some_test() {
    skip
    ....
}

@test "Test whether the files are copied, committed and pushed correctly from GitHub branch to GitLab branch if there is a file-difference." {
    ....
}
```
And then run the `./test/libs/bats/bin/bats test/test_helper_gitlab_modify.bats` file.

A short note, the test will fail because we applied the styleguide rule about local variables:
```
local somevariable
somevariable=...
```
While I had defined some global variable(s) without capitalisation, so it used to first get the value of the global `somevariable` and then remove the newline character from it, and then store that modified value back into the global. However, due to the update, I assume, the `local somevariable` overwrites the value that the global variable had to `nothing`, and then it tries to remove the newline from nothing, and store that new nothing into the global variable, which breaks the rest of the code. So to fix that we should replace all globals that I defined without capitalisation, consistently with capitalisation in all the code. So that is a learning moment for me to not ever do that again.

Reply:
https://github.com/Simple-Setup/Self-host-GitLab-Server-and-Runner-CI/issues/39

Reply:
Thanks for the explanation! I'll try the second form of testing that you have explained.